### ISSUE TITLE
No raw loops

### ISSUE THREAD

**Issue by elizabethshirley** [jsks/parmap#4]
This has become big in the C++ community, and I think it should become big in C too - basically, aside from basic setup and teardown, loops should be isolated in their own functions, and those functions should be made as generic as possible.

This important because it allows easy extension and deprecation of your programs. As they grow and evolve the "Mean-time-to-change" doesn't grow much - which is important when you're working on larger systems. The more time it takes to read, process, and act upon a change or bug typically increases dramatically, unless you constantly cleanup after yourself - and that's facilitated through generic programming. While C doesn't have templates, it does have (void*) and function pointers, which can allow easy extension. 

When you have two back to back while-loops, or nested loops, ask yourself if you really want those there, and if they wouldn't be better served by discrete functions

**Comment by jsks** [comment]
I agree with the sentiment, but wouldn't this potentially introduce a lot of unnecessary complexity for C? It makes sense in the context of attempting to create a generic algorithm that accepts a callback, but I'm not really sure how to apply it to say this project. Although I definitely agree about avoiding overly complex nested loops.

**Comment by elizabethshirley** [comment]
It's hard to say without Unit Tests (I do like the integration tests btw)

TDD reminder:
- Receive new requirement, add tests which define requirement
- Go green as quickly as possible
- Remove Duplicates, Sanitize Code Smells, Apply Design Patterns
- rinse and repeat

Definitely check out [Code As Risk by Kevlin Henney](https://www.youtube.com/watch?v=YyhfK-aBo-Y) - He does a C++ refactor later on, but it applies to C code as well.

**Comment by jsks** [comment]
Awesome, thanks for the link, I'll check it out.
